### DISCUSSION TITLE
ColumnTransform DateTime to IntColumn?

### DISCUSSION THREAD

**Discussion by madsane29** [simolus3/drift#2090] [General]
Basically I had a DateTime column in my DB that I want to change into an IntColumn since I need the milliseconds/microseconds.

I had this:
```
class MyEntity extends Table {
   DateTimeColumn? get myDate => dateTime().nullable()();
}
```

Now I have this:
```
class MyEntity extends Table {
   IntColumn? get myDate=> integer().map(const MicroDateConverter()).nullable()();
}

class MicroDateConverter extends TypeConverter<DateTime, int> {
  const MicroDateConverter();

  @override
  DateTime fromSql(int fromDb) {
    return DateTime.fromMicrosecondsSinceEpoch(fromDb);
  }

  @override
  int toSql(DateTime value) {
    return value.microsecondsSinceEpoch;
  }
}
```

I increased the schema version, wrote this migrator:

```
 if (from < 2) {
            await m.alterTable(
              TableMigration(munkanaplos, columnTransformer: {
                myentity.mydate: myentity.mydate.cast<int>(),
              }),
            );
          }
```

It runs, the problem that after the transformation the date of the existing entities turns into 1970... instead stay as 2022 after I check the value at `epochconverter.com`.

I tried it like this:

```
await m.alterTable(
              TableMigration(munkanaplos, columnTransformer: {
                myentity.mydate: FunctionCallExpression<DateTime>(
                  'strftime',
                  [
                    Constant('%s'),
                    FunctionCallExpression(
                      'datetime',
                      [
                        myentity.mydate,
                      ],
                    ),
                  ],
                ),
              }),
```

But the values end up as null.

What is the way to transform existing DateTime column into IntColumn and migrate existing values from DateTime to int?

Thanks in advance.

**Comment by madsane29** [comment]
Alright, I guess I found the answer: you don't have to do column transformation if you change from DateTimeColumn to IntColumn since in the database the column is created with the type of `INTEGER` both ways. So basically change this:

```
 DateTimeColumn? get myDate => dateTime().nullable()();
```
to 
```
IntColumn? get myDate=> integer().nullable()();
```

and you are good to go...?

**Comment by simolus3** [comment]
> Basically I had a DateTime column in my DB that I want to change into an IntColumn since I need the milliseconds/microseconds.

Please be aware that most datetime functions in sqlite only work with a second accuracy, so some transformations may still loose milliseconds if you want to use them.
If you just want to store date times with that precision and don't need to express things like "add two days" in SQL, storing them as an integer interpreted as milliseconds is fine. Otherwise, I strongly recommend storing them [as strings](https://drift.simonbinder.eu/docs/getting-started/advanced_dart_tables/#datetime-options), the documentation also provides ideas on how to migrate between the two modes. That storage option also works better with time zones.

> and you are good to go...?

If you previously stored date times as seconds since the unix epoch and are now interpreting them as milliseconds or even microseconds, you need to multiple old values in the database by 10³ or 10⁶, respectively. You could adapt your migration logic to something like this then:

```dart
await m.alterTable(
              TableMigration(munkanaplos, columnTransformer: {
                myentity.mydate: myentity.mydate.dartCast<int>() * 1000
              }),
```

**Reply by madsane29** [reply]
I just store/read the value, no change needed.

I ran into the 10^3 and 10^6 problem, that migrator might be better... I fixed it by changing my TypeConverter:

```
class MicroDateConverter extends TypeConverter<DateTime, int> {
  const MicroDateConverter();

  @override
  DateTime fromSql(int fromDb) {
    return DateTime.fromMicrosecondsSinceEpoch(_toMicroSeconds(fromDb));
  }

  @override
  int toSql(DateTime value) {
    return value.microsecondsSinceEpoch;
  }

  int _toMicroSeconds(int value) {
    while ("$value".length < 16) {
      value *= 10;
    }

    return value;
  }
}
```

Thanks for the response

**Reply by simolus3** [reply]
Looking at that converter, I hope no one will ever store unix timestamp 0 in your app :D Apart from that it should work though, you can probably replace the `while` loop with something like `if (value < 1000000000000000) value *= 1000000`.

**Reply by madsane29** [reply]
It is a value that is generated by the backend, holds the value when was the last modification on given entity... So hopefully not. Basically I needed this only to "fix" the already existing entities on the mobiles.
